### PR TITLE
mitigation for aws ec2 nvme random device name assignment + scsi device rescan + auto pvresize + skip scsi if scsi adapter not installed

### LINKED ISSUE #51: AWS nvme to sd device name support
Hi,

Thanks for this great ansible role ! What about to add support for AWS nvme device name abstraction using a temporary /dev/sd (/dev/xvd) device name ?

Example:

On new nitro based AWS EC2 instances NVMe driver is used that on EC2 startup rename defined /dev/sdX (or /dev/xvdX) device names to unpredictable /dev/nvmeXnY device names. Some solutions can be used to get /dev/sd (/dev/xvd) mapping from nvme block device definition . This role can support something like this:

1. Use nvme-cli to make symbolic links for /dev/sd or /dev/xvd from which nvmeXnY was renamed
2. Use those links in the ansible configuration to manage LVM (equal to AWS EBS configuration device_name=/dev/sdX)
3. Remove temporary created symbolic links

It'll avoid inconsistent nvme "device names" in favor of fixed /dev/sd|/dev/xvd device names and will allow this role to be used for AWS EC2 nitro based instance without add separated ansible (or provisioning like terraform) code to do that.

Some refs to find mapping between /dev/sd|/dev/xvd and /dev/nvmeXnY device names
[https://github.com/leboncoin/terraform-aws-nvme-example/blob/master/scripts/ebs_alias.sh.tpl](https://github.com/leboncoin/terraform-aws-nvme-example/blob/master/scripts/ebs_alias.sh.tpl)
[https://russell.ballestrini.net/aws-nvme-to-block-mapping/#nvme-to-block-mapping](https://russell.ballestrini.net/aws-nvme-to-block-mapping/#nvme-to-block-mapping)
[https://medium.com/@ripon.banik/how-to-find-ebs-volume-id-for-nvme-volume-1148d7499dc](https://medium.com/@ripon.banik/how-to-find-ebs-volume-id-for-nvme-volume-1148d7499dc)

### PR DESCRIPTION
**Pull request by Roxyrob**
This PR fix: #51, #46 and #58

### CODE REVIEW

File: tasks/centos.yml [hunk 1]
```diff
@@ -5,6 +5,75 @@
...
+      template:
+        src: '70-ec2-nvme-devices.rules.j2'
+        dest: '/etc/udev/rules.d/70-ec2-nvme-devices.rules'
+        mode: 0755
+      register: udev_rule
+
+    - name: centos | reload and trigger udev rules
+      shell:
+        cmd: udevadm control --reload-rules && udevadm trigger
+      when: udev_rule.changed
```
Comment (markgoddard) [inline tasks/centos.yml]: Given that this code is specific to AWS, could you move it to a separate tasks file, guarded by something like this:

```
- include_tasks: amazon.yml
  when: ansible_distribution == 'Amazon'
```

File: tasks/create_vg.yml [hunk 2]
```diff
@@ -9,6 +9,19 @@
     - vg.create is defined
     - vg.create|bool
 
+### workaround: auto pvresize waiting for upgrade to new module supporting integrated pvresize
+###             ref: https://docs.ansible.com/ansible/3/collections/community/general/lvg_module.html
+- name: create_vg | pvresize to max (implemented from ansible 2.10 and above)
```
Comment (markgoddard) [inline tasks/create_vg.yml]: Probably don't need this in the task name: ` (implemented from ansible 2.10 and above)`

**Review by markgoddard — CHANGES_REQUESTED** [review]
Thanks

File: tasks/centos.yml (hunk 1 above)
Comment (Roxyrob) [reply tasks/centos.yml]: Hi @markgoddard, in reality specific to AWS but also to Centos (for sure centos8) and nitro EC2 instances (using new nvme driver)... so many specs to manage...

But are you sure about  this ? Now it is transparent and specific to centos (for backword compatibility with debian - probably can be ported but not teste as now).

Also, ansible_distribution in your idea where come from ? a static configuration variable or set_fact in centos.yml o in other file (e.g main.tf)?

File: tasks/create_vg.yml (hunk 2 above)
Comment (Roxyrob) [reply tasks/create_vg.yml]: You're right... commited changed task comment (48e3be3c5c8432d59189ecd90ef8fa66b1fa851c)

File: tasks/centos.yml (hunk 1 above)
Comment (Roxyrob) [reply tasks/centos.yml]: Also, the code snippet is inside a conditional block. **"All block of tasks along with this code snippet"** we'll be entirely execute only for Amazon ec2 with ebs storage device using nvme driver. Not sure but this issue probably could be on other "components combination" too !?

Comment (markgoddard) [reply tasks/centos.yml]: I see, you are right - we could hit this on any distro. What we want to know is whether we are running in AWS. I believe this check will do that:

```
"ansible_system_vendor" == "Amazon EC2"
```

My concern is around performance at scale. Even skipped tasks can take a significant time to execute when there are many hosts in the inventory. Using include_tasks is dynamic, and if skipped, the included tasks will not incur any penalty.

Checkout my investigations here: https://github.com/stackhpc/ansible-scaling

Comment (Roxyrob) [reply tasks/centos.yml:75]: "Ansible scaling" seems an interesting work, I'll read more as soon as possible.

I believe that the compromise between performance and automatic decisions should be considered.

Tell me more about the check you suggested, e.g. where in your proposition in the code should be assigned "Amazon EC" value to "ansible_system_vendor" variable ? From the currents dynamic set_fact

```
- name: centos | check for nvme devices
  shell: |
    set -o pipefail
    cat /proc/partitions | awk '{print $4}' | grep -q nvme. ; echo $?
  args:
    executable: /bin/bash
  register: blkdev_nvme
  changed_when: false

- name: centos | check for amazon ebs devices
  shell: |
    set -o pipefail
    lsblk -O -J | grep -qi "amazon elastic block store" ; echo $?
  register: blkdev_awsebs
  changed_when: false
  args:
    executable: /bin/bash

- name: centos | set flag for nvme subsystem
  set_fact:
    device_is_nvme: "{{ true if (blkdev_nvme.stdout == '0') else false }}"

- name: centos | set flag for aws ebs devices
  set_fact:
    device_is_awsebs: "{{ true if (blkdev_awsebs.stdout == '0') else false }}"
```

Or from a static choise as host variables ?

```
ansible_system_vendor: 'Amazon EC2'
```

Comment (Roxyrob) [reply tasks/centos.yml]: Concerns about performance at scale mitigated in cb792e5

Comment (markgoddard) [reply tasks/centos.yml]: Perfect, thanks

Comment (Roxyrob) [reply tasks/centos.yml]: Hi @markgoddard to be consistent I need to chang correct inside amazon.yml all:
```
"- name: centos | ..."
```
to 
```
"- name: amazon | ..."
```

or
```
"- name: centos | amazon | ..."
```

what do you prefer (which is your comment standard for this) ?

Comment (markgoddard) [reply tasks/centos.yml]: Probably just amazon makes sense.

**Review by markgoddard — COMMENTED** [review]
I can't really comment on the AWS-specific code, but the rest looks fine to me.

File: tasks/centos.yml (hunk 1 above)
Comment (Roxyrob) [reply tasks/centos.yml]: Done. Can you merge this PR ?

Comment (markgoddard) [reply tasks/centos.yml]: I'm afraid I don't have the power, @mrlesmithjr does

**Comment by Roxyrob** [comment]
Hi @markgoddard, ok, waiting for PR merge by @mrlesmithjr

**Comment by Roxyrob** [comment]
Hi, any news on this merge ?

**Pull request merged**